crates/burn-jit/src/compute/kernel.rs

```diff
@@ -1,112 +1,191 @@
-    /// Launch information.
-    fn workgroup(&self) -> WorkGroup;
+/// Notes: by default, only Jit variant exists,
+/// but users can add more kernels from source by activating the
+/// template feature flag.
+pub enum Kernel {
+    /// A JIT GPU compute shader
+    JitGpu(Box<dyn JitGpuKernel>),
+    #[cfg(feature = "template")]
+    /// A kernel created from source
+    Custom(Box<dyn TemplateKernel>),
 }
 
-impl Kernel for Arc<dyn Kernel> {
-    fn source(&self) -> SourceTemplate {
-        self.as_ref().source()
+impl Kernel {
+    /// ID of the kernel, for caching
+    pub fn id(&self) -> String {
+        match self {
+            Kernel::JitGpu(shader) => shader.id(),
+            #[cfg(feature = "template")]
+            Kernel::Custom(template_kernel) => template_kernel.id(),
+        }
     }
 
-    fn id(&self) -> String {
-        self.as_ref().id()
+    /// Source of the shader, as string
+    pub fn compile(&self) -> CompiledKernel {
+        match self {
+            Kernel::JitGpu(shader) => shader.compile(),
+            #[cfg(feature = "template")]
+            Kernel::Custom(template_kernel) => template_kernel.compile(),
+        }
     }
 
-    fn workgroup(&self) -> WorkGroup {
-        self.as_ref().workgroup()
+    /// Launch information of the kernel
+    pub fn launch_information(&self) -> LaunchSettings {
+        match self {
+            Kernel::JitGpu(shader) => shader.launch_settings(),
+            #[cfg(feature = "template")]
+            Kernel::Custom(template_kernel) => template_kernel.launch_settings(),
+        }
     }
 }
 
-impl Kernel for Box<dyn Kernel> {
-    fn source(&self) -> SourceTemplate {
-        self.as_ref().source()
-    }
-
-    fn id(&self) -> String {
-        self.as_ref().id()
-    }
-
-    fn workgroup(&self) -> WorkGroup {
-        self.as_ref().workgroup()
-    }
+/// A kernel, compiled in the target language
+pub struct CompiledKernel {
+    /// Source code of the kernel
+    pub source: String,
+    /// Size of a workgroup for the compiled kernel
+    pub workgroup_size: WorkgroupSize,
 }
 
-/// Provides launch information specifying the number of work groups to be used by a compute shader.
-#[derive(new, Clone, Debug)]
-pub struct WorkGroup {
-    /// Work groups for the x axis.
-    pub x: u32,
-    /// Work groups for the y axis.
-    pub y: u32,
-    /// Work groups for the z axis.
-    pub z: u32,
+/// Informations needed to launch the kernel
+pub struct LaunchSettings {
+    /// Layout of workgroups for the kernel
+    pub workgroup: WorkGroup,
 }
 
-impl WorkGroup {
-    /// Calculate the number of invocations of a compute shader.
-    pub fn num_invocations(&self) -> usize {
-        (self.x * self.y * self.z) as usize
-    }
+/// Kernel trait with the ComputeShader that will be compiled and cached based on the
+/// provided id.
+///
+/// The kernel will be launched with the given [shader information](ShaderInformation).
+pub trait JitGpuKernel: Send + Sync {
```

Comment:
I think we can remove the Gpu here, only JitKernel. I though we would not abstract the gpu::ComputeShader, which would force the name of this trait  to have a Gpu.